Lazy date lut

DaniilAnichin opened the pull request:
<!--
Link to relevant issues or previous PRs, one per line. Use "fixes" to
automatically close an issue.
-->

- fixes #430 

<!--
Ensure each step in CONTRIBUTING.rst is complete by adding an "x" to
each box below.

If only docs were changed, these aren't relevant and can be removed.
-->

Checklist:

- [ ] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [ ] Add or update relevant docs, in the docs folder and in code.
- [ ] Ensure PR doesn't contain untouched code reformatting: spaces, etc.
- [ ] Run `flake8` and fix issues.
- [ ] Run `pytest` no tests failed. See https://clickhouse-driver.readthedocs.io/en/latest/development.html.

No checkboxes & stuff, because not finished yet. Making PR for simplification of discussion.

DaniilAnichin commented:
@xzkostyan Based on my very simplified benchmark, warming up cache for the entire lookup table is worst-case-scenario 10 times slower than simple fetch from dict; and while it's 10 times - it's still at ~50 milliseconds for full warm-up vs 5 for each access.
Can go back to env-based solution too though
```
In [107]: from datetime import date, timedelta
time: 266 µs (started: 2024-04-04 22:50:22 +03:00)

In [108]: from clickhouse_driver.columns import datecolumn
time: 394 µs (started: 2024-04-04 22:50:27 +03:00)

In [109]: epoch_start = date(1970, 1, 1)
time: 461 µs (started: 2024-04-04 22:50:15 +03:00)

In [110]: full = list(range(65536))
time: 1.17 ms (started: 2024-04-04 22:47:19 +03:00)

In [111]: def get_date_lut(item):
     ...:     return epoch_start + timedelta(days=item)
     ...: 
time: 517 µs (started: 2024-04-04 22:47:26 +03:00)

In [112]: %timeit -n 100 tuple(datecolumn.DateColumn.date_lut[i] for i in full)
4.7 ms ± 140 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
time: 3.29 s (started: 2024-04-04 22:47:33 +03:00)

In [113]: %timeit -n 100 tuple(get_date_lut(i) for i in full)
43.9 ms ± 17.3 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
time: 30.7 s (started: 2024-04-04 22:47:46 +03:00)
```

xzkostyan commented on clickhouse_driver/columns/datecolumn.py:
```diff
@@ -7,7 +7,22 @@
 
 epoch_start_date32 = date(1900, 1, 1)
-epoch_end_date32 = date(2283, 11, 11)
+# Why was it 17 year earlier that limit set in clickhouse?
```
It seems that some old version has this limitation https://github.com/mymarilyn/clickhouse-driver/actions/runs/8805937959

xzkostyan commented:
Will you finish this PR?

DaniilAnichin commented:
Fixed flake8, updated tests to reflect expected `epoch_end_date32` changes (old CH version indeed has different range, but we seem to already have different tests for that very reason)

Will be happy to finish PR) 

Anything to take care of, assuming pipelines are fine?

DaniilAnichin commented:
Well, pipelines failed on something seemingly not related to these changes; 
Anything to update in PR to make it ready to merge?

DaniilAnichin commented:
Valgrind related [issue](https://github.com/orgs/community/discussions/120966):

DaniilAnichin commented:
Valgrind issue fixed, forked pipeline [passed](https://github.com/DaniilAnichin/clickhouse-driver/actions/runs/8813014930)

xzkostyan commented:
LGTM, but what about ENV flag for lazy LUT? I don't see any changes in "Changes", but I see changes in commits.

DaniilAnichin commented:
Sorry, I made it not obvious and more convoluted than needed; 
There were initially 3 commits, with slightly different implementations, from more complicated (with range validation & option to revert to current warm lut) to less complicated (the one in last commit, with lasy lut only).

Here is how it would look like if we add env option to revert to previous implementation: [diff](https://github.com/DaniilAnichin/clickhouse-driver/compare/feature/lazy-date-lut...DaniilAnichin:clickhouse-driver:feature/lazy-date-lut-non-breaking?expand=1)
If we want to proceed with that, I'll merge it into current branch

xzkostyan commented:
Let's try with this diff.

Maybe also split `CLICKHOUSE_DRIVER_LASY_DATE_LUT` into (epoch_end, epoch_start) for building small map to aviod pre-warm?

DaniilAnichin commented:
Sounds reasonable, added proposed change, made implementation a bit more streamlined.
Should I try updating changelog & wherever we should put documentation explaining new configuration option usage, or you will rather do that closer towards release?

xzkostyan commented:
> Should I try updating changelog & wherever we should put documentation explaining new configuration option usage, or you will rather do that closer towards release?

Feel free to update docs and examples. I'll bump package version without release and there will be unreleased one in the docs.

ghazi-git commented:
Thank you @DaniilAnichin for your work on this PR and @xzkostyan for reviewing and merging it.
I was profiling the startup of a project I'm working on when I found out that the 3 dict comprehensions were near the top in total time (~200ms combined)
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       ...
        1    0.112    0.112    0.112    0.112 /home/<user>/venv/lib/python3.10/site-packages/clickhouse_driver/columns/datecolumn.py:65(<dictcomp>)
       ...
        1    0.056    0.056    0.056    0.056 /home/<user>/venv/lib/python3.10/site-packages/clickhouse_driver/columns/datecolumn.py:22(<dictcomp>)
       ...
        1    0.035    0.035    0.035    0.035 /home/<user>/venv/lib/python3.10/site-packages/clickhouse_driver/columns/datecolumn.py:69(<dictcomp>)
```

Looking forward to the next version being released

xzkostyan commented:
Version 0.2.8 is released.
